pyne/bins.pyx

```diff
@@ -118,19 +118,66 @@ def pointwise_linear_collapse(np.ndarray[np.float64_t, ndim=1] x_g,
     y_g : np.ndarray
         The group collapsed data, length G. 
     """
+    return pointwise_collapse(x_g, x, y, (False, False))
+
+@cython.boundscheck(False)
+def pointwise_collapse(np.ndarray[np.float64_t, ndim=1] x_g, 
+                       np.ndarray[np.float64_t, ndim=1] x, 
+                       np.ndarray[np.float64_t, ndim=1] y,
+                       log=(False, False)):
+    """Collapses pointwise data to G groups based on a interpolation
+    between the points. This is useful for collapsing cross section data.
+
+    Parameters
+    ----------
+    x_g : array-like
+        Group boundaries, length G+1 for G groups, must be monotonic in the 
+        same direction as x.
+    x : array-like
+        Pointwise abscissa to be collapsed, must be monotonic in the same direction
+        as x_g and have the same length as y.
+    y : array-like
+        Pointwise data to be interpolated, must have the same length as x.
+    log : tuple of bools, optional, default = (False, False)
+       Specify logrithmic interpolation in the x and y dimension, respectively
+
+    Returns
+    -------
+    y_g : np.ndarray
+        The group collapsed data, length G. 
+    """
     cdef int G = x_g.shape[0] - 1
     cdef int N = x.shape[0] - 1
     cdef int g0, g1  # current group index
     cdef int n0, n1  # current point index
     cdef double val, ylower, yupper
     cdef np.ndarray[np.float64_t, ndim=1] y_g = np.empty(G, dtype='float64')
+
+    # Ensure input is monotonically increasing/decreasing
+    for z in [x_g, x]:
+        if not (np.all(np.diff(z) > 0.) or np.all(np.diff(z) < 0)):
+            raise ValueError("x and x_g arrays must be monotonically"  
+                             "increasing/decreasing.")
+
     reversed = False
     if x_g[0] > x_g[-1]:
-        # monotonically decreaing, make increasing so logic is simpler
+        # monotonically decreasing, make increasing so logic is simpler
         x_g = x_g[::-1]
         x = x[::-1]
         y = y[::-1]
         reversed = True
+
+    # Handle logrithmic interpolations
+    if log[0]:
+        if x_g[0] <= 0.0 or x_g[0] <= 0.0:
+            raise ValueError("x values must be positive for logrithmic interpolation")
+        x_g = np.array([np.log(z) for z in x_g])
```

Comment:
This is really inefficient to have a list comprehension here.  Just use the numpy function:

```python
x_g = np.log(x_g)
```

Same for all below.